### ISSUE TITLE
[PG] Collective issue blocks the host outside CUDA graph capture, breaking `async_op=True` (regression in v0.3.13)

### ISSUE THREAD

**Issue by QiaoK** [kvcache-ai/Mooncake#3765]
## Summary

`MooncakeWorker::putTaskCuda` synchronizes the host on the non-capturing path. Because a
Mooncake collective installs a completion wait on the issue stream, the
`cudaStreamSynchronize(issue_stream)` at the top of the *next* collective blocks the calling
thread until the *previous* collective has completed.

The effect is that c10d's `async_op=True` is no longer asynchronous when the Mooncake process
group backend is used. `dist.all_reduce(..., async_op=True)` still returns a `Work` handle
immediately, but the following collective will not be issued until the earlier one has
finished, so the asynchrony the handle advertises does not exist.

Introduced by 4f9e6749 ("fix(ep,pg): prevent cudaErrorIllegalAddress during CUDA graph
capture with TBO", #3609, 2026-08-25). The first release containing it is **v0.3.13**;
v0.3.12.post1 is unaffected.

## Where

`mooncake-pg/src/mooncake_worker_host.cpp`, in `putTaskCuda`. Before #3609 the two streams
were ordered with an event in all cases. After it, the non-capturing path instead does:

```cpp
if (is_capturing) {
    GpuEvent event_start(issue_stream.deviceIndex());
    event_start.record(issue_stream);
    enq_stream.waitEvent(event_start);
} else {
    // Do not create an eager cross-stream cycle with the completion wait
    // installed by the preceding collective on issue_stream.
    PG_ASSERT_CUDA(cudaStreamSynchronize(issue_stream.get()));
}
```

plus two further `cudaStreamSynchronize` calls later in the same function. The comment
identifies the mechanism precisely: the preceding collective has installed a completion wait
on `issue_stream`, so synchronizing that stream waits for the preceding collective to finish.

Worth noting that the illegal-address problem #3609 set out to fix concerns the *capturing*
path, but the behavioural change lands on the non-capturing path, which is the one every
eager-mode caller uses.

### 1. Communication/computation overlap is silently lost

The standard reason to pass `async_op=True` is to overlap communication with unrelated work:

```python
handle = dist.all_reduce(grad, async_op=True)
loss = compute_something_else()   # expected to run while the all_reduce is in flight
handle.wait()
```

This still produces correct results, but the overlap is gone as soon as the interleaved work
contains another collective, and the cost is invisible: no warning, no API change, just a
throughput regression that profiles as time spent inside the *next* collective rather than in
the one that is actually being waited on.

### 2. Any group whose ranks are not in lockstep can deadlock

This is the more serious consequence. Consider three ranks where one is briefly ahead, which
is normal in any application with data-dependent control flow, uneven batches, or a
coordination step that one rank initiates:

1. Rank A posts an async collective and, per the c10d contract, continues without waiting.
2. Rank A reaches its next collective. Issuing it calls `cudaStreamSynchronize(issue_stream)`,
   which now waits for the collective from step 1 to complete.
3. That collective cannot complete until ranks B and C join it.
4. Ranks B and C are still at the earlier collective, waiting for rank A — which never
   enqueues it, because it is blocked in step 2.

Nothing times out and no Mooncake error is raised; the group simply stops. Before #3609 rank A
would have issued its next collective without blocking, B and C would have completed it, and
all three would have converged.

The pattern that triggers this — post an async collective, then keep participating in other
collectives while it is outstanding — is exactly what `async_op=True` exists to support. It
covers non-blocking barriers polled across iterations, background/bucketed gradient
reductions, and any coordination where one rank proposes and the others notice later. It also
means a group can hang while every rank is inside a perfectly ordinary collective call, which
makes the failure very hard to attribute: a stack trace shows only the *later* collective,
not the earlier one whose completion is actually being awaited.

## Reproduction and measurement

The deadlock needs two ingredients: an async collective left outstanding, and another
collective issued on the same group before it completes. Sketch:

```python
# every rank
if rank == 0:
    handle = dist.barrier(group=g, async_op=True)   # returns immediately
    dist.all_reduce(t, group=g)                     # blocks here on the barrier
    handle.wait()
else:
    time.sleep(1)                                   # stand-in for any skew
    dist.all_reduce(t, group=g)                     # never satisfied: rank 0 never enqueues
    dist.barrier(group=g)
```

Measured on a workload that uses this pattern, with only the Mooncake build varying (same
torch, same driver, same application code), 88 iterations per configuration:

| Mooncake | result |
|---|---|
| `3b5a594` (before #3609) | 88/88 pass |
| `a11f54e` (main, after #3609) | 72/88 pass, 16 hangs |

Every failure has the same signature: the group stalls with 2 of 5 ranks arrived and makes no
further progress. Narrowing to the four configurations that fail most often, 5 iterations each:

| Mooncake | result |
|---|---|
| `a11f54e` unmodified | 10/20 pass |
| `a11f54e` with the non-capturing path reverted to event ordering | 20/20 pass |
| `3b5a594` | 20/20 pass |

## Suggested fix

Restore the event-based ordering on both paths and drop the three host synchronizations,
keeping the improved capture detection that #3609 added so the CUDA-graph case it targets
still takes the path it does today:

```cpp
    const bool is_capturing = issue_capture != cudaStreamCaptureStatusNone ||
                              enq_capture != cudaStreamCaptureStatusNone;
    GpuEvent event_start(issue_stream.deviceIndex());
    event_start.record(issue_stream);
    enq_stream.waitEvent(event_start);
    ...
    if (!is_capturing) {
        waitUntilTasksSubmitted(submitted_tasks);
    }

    GpuEvent event_end(enq_stream.deviceIndex());
    event_end.record(enq_stream);
    issue_stream.waitEvent(event_end);
```

This is the variant measured as 20/20 above. If the eager cross-stream cycle the comment
refers to is a real hazard for TBO, it would be better to gate the host synchronization on
the specific case that needs it rather than applying it to every eager collective, since as
written it changes the documented semantics of `async_op=True` for all callers.

## Environment

- Mooncake `a11f54e5cc` (main) and `3b5a594`, built from source, CUDA 13, `manylinux_2_39`
- 5-rank single-node and multi-node process groups over the Mooncake PG backend
- torch 2.13.0+cu130

**Comment by 200lz** [comment]
I traced this one layer further into enqueueTaskKernel, and I think there is an additional constraint that matters before implementing the suggested eager-path revert.

enqueueTaskKernel is a blocking device-side spin kernel: after publishing the task it stays active until the worker has completed the transfer/sync protocol. Today all CUDA task slots on a device share the single MooncakeWorker::enqueue_stream_.

That means simply removing the three host-side cudaStreamSynchronize calls does not fully restore asynchronous issuing across consecutive collectives. Collective N+1’s enqueue kernel is still queued behind N’s spin kernel on the same enqueue stream, so waitUntilTasksSubmitted for N+1 can still become a completion wait for N.

This also changes how I would reproduce the regression: the clean measurement is the issue latency of the first async collective under rank skew. On current main, the issuing call itself waits for completion; pre-#3609 it returns promptly. The two-collective reproducer in the issue appears to rely on mismatched collective ordering across ranks, so I would avoid using that as the correctness test.

The smallest structural fix I see is:

keep #3609’s improved capture detection;
restore event ordering on the eager path and remove the host synchronizations;
give each CUDA task slot its own enqueue stream, so one slot’s spin kernel cannot serialize another slot;
keep the existing submitted-sequence gate for now.

This also seems consistent with the current comment saying “one enqueue stream per task slot”, even though the implementation still has a single shared enqueue_stream_.

I’d be happy to prototype this direction and add the corrected eager async regression test if this matches the intended follow-up to #3609. I would keep the existing multi-communicator/task-slot accounting issues out of the first PR unless they prove necessary.

The one validation gap on my side is the original TBO/CUDA-graph case from #3609; if the patch shape looks right, I can do the eager-path validation and would appreciate help validating the original H20/TBO graph scenario as well.

**Comment by QiaoK** [comment]
Thanks for digging into this so carefully — your tracing of `enqueueTaskKernel` and the shared
enqueue stream went a good deal deeper than my original report, and I appreciate it.

I tried to back the report up with a small standalone reproducer (`pg_test_utils.py` style,
5 ranks, single node, arrival skew) and I could not reproduce the behaviour I described.
Issuing an async `all_reduce` under skew takes essentially the same time on main, on pre-#3609,
and on the reverted build — about 5 s in all three — so the claim that #3609 changed
`async_op=True` semantics isn't supported. In the one pattern where the builds do differ, a data
collective issued while an async barrier is still outstanding, main is the fast one (~4 ms
versus ~5 s). That suggests the eager-path revert I proposed is the wrong direction, so please
don't spend time on it on my account.

All of which points at something specific to my setup rather than a general PG problem, so I'll
close this. If I manage to pin it down to something reproducible, I'll open a fresh issue built
around the reproducer.

Two points from your comment look worth keeping on their own merits, and I verified both against
`a11f54e`: `enqueue_stream_` is a single stream on the worker
(`mooncake-pg/include/mooncake_worker.cuh:140`), so the comment describing one enqueue stream
per task slot doesn't match the implementation; and `enqueueTaskKernel` blocks for the lifetime
of the task (`mooncake-pg/src/mooncake_worker.cu:38-41`).

Thanks again for the help here.
